node/subsystem-util/src/rolling_session_window.rs

```diff
@@ -146,19 +158,105 @@ impl RollingSessionWindow {
+			sender.send_message(ChainApiMessage::FinalizedBlockNumber(tx)).await;
+			match rx.await {
+				Ok(Ok(number)) => number,
+				Ok(Err(e)) =>
+					return Err(SessionsUnavailable {
+						kind: SessionsUnavailableReason::ChainApi(e),
+						info: None,
+					}),
+				Err(err) => {
+					gum::warn!(
+						target: LOG_TARGET,
+						?err,
+						"Failed fetching last finalized block number"
+					);
+					return Err(SessionsUnavailable {
+						kind: SessionsUnavailableReason::MissingLastFinalizedBlock,
+						info: None,
+					})
+				},
+			}
+		};
+
+		let (tx, rx) = oneshot::channel();
+		// We want to get the parent of the last finalized block.
+		sender
+			.send_message(ChainApiMessage::FinalizedBlockHash(
+				last_finalized_height.saturating_sub(1),
+				tx,
+			))
+			.await;
+		let last_finalized_hash_parent = match rx.await {
+			Ok(Ok(maybe_hash)) => maybe_hash,
+			Ok(Err(e)) =>
+				return Err(SessionsUnavailable {
+					kind: SessionsUnavailableReason::ChainApi(e),
+					info: None,
+				}),
+			Err(err) => {
+				gum::warn!(target: LOG_TARGET, ?err, "Failed fetching last finalized block hash");
+				return Err(SessionsUnavailable {
+					kind: SessionsUnavailableReason::MissingLastFinalizedBlockHash,
+					info: None,
+				})
+			},
+		};
+
+		// Get the session in which the last finalized block was authored.
+		if let Some(last_finalized_hash_parent) = last_finalized_hash_parent {
+			let session =
+				match get_session_index_for_child(sender, last_finalized_hash_parent).await {
+					Ok(session_index) => session_index,
+					Err(err) => {
+						gum::warn!(
+							target: LOG_TARGET,
+							?err,
+							?last_finalized_hash_parent,
+							"Failed fetching session index"
+						);
+						return Err(err)
+					},
+				};
+
+			Ok(session)
+		} else {
+			return Err(SessionsUnavailable {
+				kind: SessionsUnavailableReason::MissingLastFinalizedBlockHash,
+				info: None,
+			})
+		}
+	}
+
 	/// When inspecting a new import notification, updates the session info cache to match
 	/// the session of the imported block's child.
 	///
 	/// this only needs to be called on heads where we are directly notified about import, as sessions do
 	/// not change often and import notifications are expected to be typically increasing in session number.
 	///
 	/// some backwards drift in session index is acceptable.
-	pub async fn cache_session_info_for_head(
+	pub async fn cache_session_info_for_head<Sender>(
 		&mut self,
-		sender: &mut impl overseer::SubsystemSender<RuntimeApiMessage>,
+		sender: &mut Sender,
 		block_hash: Hash,
-	) -> Result<SessionWindowUpdate, SessionsUnavailable> {
+	) -> Result<SessionWindowUpdate, SessionsUnavailable>
+	where
+		Sender: overseer::SubsystemSender<RuntimeApiMessage>
+			+ overseer::SubsystemSender<ChainApiMessage>,
+	{
 		let session_index = get_session_index_for_child(sender, block_hash).await?;
+		let last_finalized_block_session = self.last_finalized_block_session(sender).await?;
```

Comment:
Technically it is the earliest_non_finalized_block_session. (Which is exactly what we want)